pkg/controller/containerrecreaterequest/crr_controller.go

```diff
@@ -0,0 +1,321 @@
+	start := time.Now()
+	klog.V(3).Infof("Starting to process CRR %v", request.NamespacedName)
+	defer func() {
+		if err != nil {
+			klog.Warningf("Failed to process CRR %v, elapsedTime %v, error: %v", request.NamespacedName, time.Since(start), err)
+		} else if res.RequeueAfter > 0 {
+			klog.Infof("Finish to process CRR %v, elapsedTime %v, RetryAfter %v", request.NamespacedName, time.Since(start), res.RequeueAfter)
+		} else {
+			klog.Infof("Finish to process CRR %v, elapsedTime %v", request.NamespacedName, time.Since(start))
+		}
+	}()
+
+	crr := &appsv1alpha1.ContainerRecreateRequest{}
+	err = r.Get(context.TODO(), request.NamespacedName, crr)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	pod := &v1.Pod{}
+	podErr := r.Get(context.TODO(), types.NamespacedName{Namespace: crr.Namespace, Name: crr.Spec.PodName}, pod)
+	if podErr != nil && !errors.IsNotFound(podErr) {
+		return reconcile.Result{}, fmt.Errorf("failed to get Pod for CRR: %v", podErr)
+	}
+	isPodDeleted := errors.IsNotFound(podErr) || pod.DeletionTimestamp != nil || string(pod.UID) != crr.Labels[appsv1alpha1.ContainerRecreateRequestPodUIDKey]
+
+	if (crr.DeletionTimestamp != nil || crr.Status.CompletionTime != nil) && !isPodDeleted &&
+		slice.ContainsString(crr.Finalizers, appsv1alpha1.ContainerRecreateRequestUnreadyAcquiredKey, nil) {
+		if err := r.releasePodNotReady(crr, pod); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	if crr.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	if crr.Status.CompletionTime != nil {
+		if _, ok := crr.Labels[appsv1alpha1.ContainerRecreateRequestActiveKey]; ok {
+			body := fmt.Sprintf(`{"metadata":{"labels":{"%s":null}}}`, appsv1alpha1.ContainerRecreateRequestActiveKey)
+			return reconcile.Result{}, r.Patch(context.TODO(), crr, client.RawPatch(types.MergePatchType, []byte(body)))
+		}
+
+		var leftTime time.Duration
+		if crr.Spec.TTLSecondsAfterFinished != nil {
+			leftTime = time.Duration(*crr.Spec.TTLSecondsAfterFinished)*time.Second - time.Since(crr.Status.CompletionTime.Time)
+			if leftTime <= 0 {
+				klog.Infof("Deleting CRR %s/%s for ttlSecondsAfterFinished", crr.Namespace, crr.Name)
+				if err = r.Delete(context.TODO(), crr); err != nil {
+					return reconcile.Result{}, fmt.Errorf("delete CRR error: %v", err)
+				}
+				return reconcile.Result{}, nil
+			}
+		}
+		return reconcile.Result{RequeueAfter: leftTime}, nil
+	}
+
+	if isPodDeleted {
+		klog.Warningf("Complete CRR %s/%s as failure for Pod %s with UID=%s has gone",
+			crr.Namespace, crr.Name, crr.Spec.PodName, crr.Labels[appsv1alpha1.ContainerRecreateRequestPodUIDKey])
+		return reconcile.Result{}, r.completeCRR(crr, "pod has gone")
```

Comment:
after completeCRR, shall we RequeueAfter some duration according to TTLSecondsAfterFinished?

Reply:
after completeCRR, the new CRR will be reconciled again and requeueAfter at that moment